File: browser/src/pipeline.js
import {
  ENTITY_PRIORITY,
  MANUAL_ENTITY_TYPES,
  MAX_TEXT_LENGTH,
  PLACEHOLDER_LABELS,
  POLICY_ID,
  normalizeEntityType,
} from "./policy.js";
import { detectWithRules, isInsidePlaceholder } from "./rules.js";

const CONTEXTUAL_TYPE_RECOGNIZERS = Object.freeze({
  LOCATION: new Set([
    "concatenated-facility-name",
    "contextual-facility-name",
    "labeled-facility-name",
  ]),
  PERSON: new Set(["labeled-name", "relationship-name", "titled-clinician-name"]),
  UNIQUE_ID: new Set(["other-contextual-identifier", "provider-identifier"]),
});

function detectionRank(detection) {
  const contextualRecognizers = CONTEXTUAL_TYPE_RECOGNIZERS[detection.entityType];
  const contextual = contextualRecognizers
    ? detection.recognizers.some((recognizer) => contextualRecognizers.has(recognizer))
    : false;
  return [Number(contextual), ENTITY_PRIORITY[detection.entityType] ?? 0, detection.score];
}

function isHigherRank(candidateRank, currentRank) {
  return candidateRank.some(
    (value, index) =>
      value > currentRank[index] &&
      candidateRank.slice(0, index).every((earlier, earlierIndex) => earlier === currentRank[earlierIndex]),
  );
}

function validateInput(text, manualFindings) {
  if (typeof text !== "string" || !text.trim()) throw new Error("Text is required.");
  if (text.length > MAX_TEXT_LENGTH) throw new Error("Text exceeds the local processing limit.");
  if (text.includes("\0")) throw new Error("Text contains an unsupported null character.");
  if (!Array.isArray(manualFindings) || manualFindings.length > 1_000) {
    throw new Error("Manual review data is invalid.");
  }
  for (const finding of manualFindings) {
    if (!MANUAL_ENTITY_TYPES.includes(finding.entity_type)) {
      throw new Error("Manual review data is invalid.");
    }
    if (
      !Number.isInteger(finding.start) ||
      !Number.isInteger(finding.end) ||
      finding.start < 0 ||
      finding.end <= finding.start ||
      finding.end > text.length
    ) {
      throw new Error("Manual review data is invalid.");
    }
  }
}

export function mergeOverlaps(detections) {
  const ordered = [...detections].sort(
    (left, right) =>
      left.start - right.start ||
      left.end - right.end ||
      right.score - left.score ||
      left.entityType.localeCompare(right.entityType),
  );
  if (!ordered.length) return [];

  const merged = [];
  let current = ordered[0];
  for (const candidate of ordered.slice(1)) {
    if (candidate.start >= current.end) {
      merged.push(current);
      current = candidate;
      continue;
    }

    const currentRank = detectionRank(current);
    const candidateRank = detectionRank(candidate);
    const candidateWins = isHigherRank(candidateRank, currentRank);
    current = {
      start: Math.min(current.start, candidate.start),
      end: Math.max(current.end, candidate.end),
      entityType: candidateWins ? candidate.entityType : current.entityType,
      score: Math.max(current.score, candidate.score),
      recognizers: [...new Set([...current.recognizers, ...candidate.recognizers])].sort(),
    };
  }
  merged.push(current);
  return merged;
}

function replaceDetections(text, detections) {
  const counters = new Map();
  const valueTokens = new Map();
  const replacements = [];

  for (const detection of detections) {
    const label = PLACEHOLDER_LABELS[normalizeEntityType(detection.entityType)];
    const valueKey = `${label}\0${text.slice(detection.start, detection.end).toLocaleLowerCase()}`;
    let replacementValue = valueTokens.get(valueKey);
    if (!replacementValue) {
      const counter = (counters.get(label) ?? 0) + 1;
      counters.set(label, counter);
      replacementValue = `[${label}_${counter}]`;
      valueTokens.set(valueKey, replacementValue);
    }
    replacements.push({ start: detection.start, end: detection.end, value: replacementValue });
  }

  let cleanedText = text;
  for (const replacement of [...replacements].reverse()) {
    cleanedText =
      cleanedText.slice(0, replacement.start) +
      replacement.value +
      cleanedText.slice(replacement.end);
  }
  return { cleanedText, replacements };
}

function mapOriginalSpan(start, end, replacements) {
  let delta = 0;
  for (const replacement of replacements) {
    if (replacement.end <= start) {
      delta += replacement.value.length - (replacement.end - replacement.start);
      continue;
    }
    if (replacement.start >= end) break;
    return null;
  }
  return [start + delta, end + delta];
}

function mapExcludedSpans(findings, replacements) {
  const exemptions = [];
  for (const finding of findings) {
    if (finding.selected) continue;
    const mapped = mapOriginalSpan(finding.start, finding.end, replacements);
    if (mapped) exemptions.push([mapped[0], mapped[1], finding.entity_type]);
  }
  return exemptions;
}

function isReviewedExemption(detection, exemptions) {
  return exemptions.some(
    ([start, end, entityType]) =>
      detection.start >= start && detection.end <= end && detection.entityType === entityType,
  );
}

async function detectAll(text, nerDetector, progressCallback) {
  const detections = detectWithRules(text);
  if (nerDetector?.ready) {
    detections.push(...(await nerDetector.detect(text, progressCallback)));
  }
  return detections.filter(
    (detection) =>
      detection.start >= 0 &&
      detection.end > detection.start &&
      detection.end <= text.length &&
      !isInsidePlaceholder(text, detection),
  );
}

function publicFinding(detection, findingId, selected, source = "automatic") {
  return {
    finding_id: findingId,
    start: detection.start,
    end: detection.end,
    entity_type: normalizeEntityType(detection.entityType),
    score: Number(detection.score.toFixed(4)),
    recognizers: detection.recognizers,
    selected,
    source,
  };
}

export async function deidentify({
  text,
  excludedFindingIds = [],
  manualFindings = [],
  nerDetector = null,
  progressCallback = () => {},
}) {
  validateInput(text, manualFindings);
  const exclusions = new Set(excludedFindingIds);
  const automaticDetections = mergeOverlaps(
    await detectAll(text, nerDetector, (progress) => progressCallback(progress * 0.5)),
  );
  const automaticFindings = automaticDetections.map((detection, index) => {
    const findingId = `auto-${String(index + 1).padStart(4, "0")}`;
    return publicFinding(detection, findingId, !exclusions.has(findingId));
  });

  const selectedDetections = automaticDetections.filter(
    (_detection, index) => automaticFindings[index].selected,
  );
  const manualDetections = manualFindings.map((finding) => ({
    start: finding.start,
    end: finding.end,
    entityType: normalizeEntityType(finding.entity_type),
    score: 1,
    recognizers: ["human-review"],
  }));
  const manualPublicFindings = manualDetections.map((detection, index) =>
    publicFinding(detection, `manual-${String(index + 1).padStart(4, "0")}`, true, "manual"),
  );

  const applied = mergeOverlaps([...selectedDetections, ...manualDetections]);
  const { cleanedText, replacements } = replaceDetections(text, applied);
  const exemptions = mapExcludedSpans(automaticFindings, replacements);
  let residualDetections = mergeOverlaps(
    await detectAll(cleanedText, nerDetector, (progress) => progressCallback(0.5 + progress * 0.5)),
  );
  residualDetections = residualDetections.filter(
    (detection) =>
      !isInsidePlaceholder(cleanedText, detection) &&
      !isReviewedExemption(detection, exemptions),
  );
  const residualFindings = residualDetections.map((detection, index) =>
    publicFinding(detection, `residual-${String(index + 1).padStart(4, "0")}`, false, "residual"),
  );

  const detectorStatuses = [
    {
      name: "deterministic-rules",
      ready: true,
      required: true,
      detail: "Browser deterministic recognizers ready",
    },
    {
      name: "browser-local-ner",
      ready: Boolean(nerDetector?.ready),
      required: true,
      detail: nerDetector?.ready
        ? "Pinned DistilBERT ONNX model ready in browser memory"
        : "Required browser NER model is unavailable; export is blocked",
    },
  ];
  const blockReasons = [];
  if (detectorStatuses.some((status) => status.required && !status.ready)) {
    blockReasons.push("A required local detector is unavailable.");
  }
  if (residualFindings.length) {
    blockReasons.push("The cleaned text still contains unresolved findings.");
  }

  const findings = [...automaticFindings, ...manualPublicFindings];
  return {
    cleaned_text: cleanedText,
    findings,
    residual_findings: residualFindings,
    detector_statuses: detectorStatuses,
    policy_id: POLICY_ID,
    applied_count: findings.filter((finding) => finding.selected).length,
    export_allowed: blockReasons.length === 0,
    export_block_reasons: blockReasons,
    manual_entity_types: MANUAL_ENTITY_TYPES,
  };
}

File: src/med_data_cleaner/deid/pipeline.py
from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterable, Sequence
from dataclasses import dataclass

from med_data_cleaner.deid.models import (
    DeidentificationResult,
    Detection,
    Detector,
    DetectorStatus,
    Finding,
    ManualFinding,
    PipelineInputError,
)
from med_data_cleaner.deid.policy import (
    ALLOWED_MANUAL_ENTITY_TYPES,
    ENTITY_PRIORITY,
    PLACEHOLDER_LABELS,
    POLICY_ID,
    normalize_entity_type,
)
from med_data_cleaner.deid.presidio_detector import PresidioDetector
from med_data_cleaner.deid.regex_detector import PLACEHOLDER_PATTERN, RegexDetector

MAX_TEXT_LENGTH = 500_000
CONTEXTUAL_TYPE_RECOGNIZERS = {
    "LOCATION": frozenset({"concatenated-facility-name", "labeled-facility-name"}),
    "PERSON": frozenset({"labeled-name", "relationship-name", "titled-clinician-name"}),
    "UNIQUE_ID": frozenset({"other-contextual-identifier", "provider-identifier"}),
}


def _detection_rank(item: Detection) -> tuple[bool, int, float]:
    contextual = CONTEXTUAL_TYPE_RECOGNIZERS.get(item.entity_type, frozenset())
    return (
        any(recognizer in contextual for recognizer in item.recognizers),
        ENTITY_PRIORITY.get(item.entity_type, 0),
        item.score,
    )


@dataclass(frozen=True, slots=True)
class Replacement:
    start: int
    end: int
    value: str


class DeidentificationPipeline:
    def __init__(
        self,
        detectors: Sequence[Detector] | None = None,
        *,
        max_text_length: int = MAX_TEXT_LENGTH,
    ) -> None:
        self.detectors: tuple[Detector, ...] = tuple(
            detectors if detectors is not None else (RegexDetector(), PresidioDetector())
        )
        self.max_text_length = max_text_length

    @property
    def detector_statuses(self) -> tuple[DetectorStatus, ...]:
        return tuple(detector.status for detector in self.detectors)

    @property
    def engine_ready(self) -> bool:
        return all(status.ready for status in self.detector_statuses if status.required)

    def deidentify(
        self,
        text: str,
        *,
        excluded_finding_ids: Iterable[str] = (),
        manual_findings: Iterable[ManualFinding] = (),
    ) -> DeidentificationResult:
        self._validate_text(text)
        exclusions = frozenset(excluded_finding_ids)
        manual = tuple(manual_findings)
        self._validate_manual_findings(text, manual)

        automatic_detections = self._merge_overlaps(self._detect_all(text))
        automatic_findings = tuple(
            Finding(
                finding_id=f"auto-{index:04d}",
                start=detection.start,
                end=detection.end,
                entity_type=detection.entity_type,
                score=detection.score,
                recognizers=detection.recognizers,
                selected=f"auto-{index:04d}" not in exclusions,
            )
            for index, detection in enumerate(automatic_detections, start=1)
        )

        selected_detections = [
            detection
            for detection, finding in zip(automatic_detections, automatic_findings, strict=True)
            if finding.selected
        ]
        manual_detections = [
            Detection(
                start=finding.start,
                end=finding.end,
                entity_type=normalize_entity_type(finding.entity_type),
                score=1.0,
                recognizers=("human-review",),
                explanation="Added during local human review",
            )
            for finding in manual
        ]
        manual_public_findings = tuple(
            Finding(
                finding_id=f"manual-{index:04d}",
                start=detection.start,
                end=detection.end,
                entity_type=detection.entity_type,
                score=detection.score,
                recognizers=detection.recognizers,
                selected=True,
                source="manual",
            )
            for index, detection in enumerate(manual_detections, start=1)
        )

        applied = self._merge_overlaps([*selected_detections, *manual_detections])
        cleaned_text, replacements = self._replace(text, applied)

        exemptions = self._map_excluded_spans(
            automatic_findings=automatic_findings,
            replacements=replacements,
        )
        residual_detections = self._merge_overlaps(self._detect_all(cleaned_text))
        residual_detections = [
            detection
            for detection in residual_detections
            if not self._inside_placeholder(cleaned_text, detection)
            and not self._is_reviewed_exemption(detection, exemptions)
        ]
        residual_findings = tuple(
            Finding(
                finding_id=f"residual-{index:04d}",
                start=detection.start,
                end=detection.end,
                entity_type=detection.entity_type,
                score=detection.score,
                recognizers=detection.recognizers,
                selected=False,
                source="residual",
            )
            for index, detection in enumerate(residual_detections, start=1)
        )

        block_reasons: list[str] = []
        if not self.engine_ready:
            block_reasons.append("A required local detector is unavailable.")
        if residual_findings:
            block_reasons.append("The cleaned text still contains unresolved findings.")

        return DeidentificationResult(
            cleaned_text=cleaned_text,
            findings=(*automatic_findings, *manual_public_findings),
            residual_findings=residual_findings,
            detector_statuses=self.detector_statuses,
            policy_id=POLICY_ID,
            export_allowed=not block_reasons,
            export_block_reasons=tuple(block_reasons),
        )

    def _validate_text(self, text: str) -> None:
        if not isinstance(text, str) or not text.strip():
            raise PipelineInputError("Text is required")
        if len(text) > self.max_text_length:
            raise PipelineInputError("Text exceeds the local processing limit")
        if "\x00" in text:
            raise PipelineInputError("Text contains an unsupported null character")

    @staticmethod
    def _validate_manual_findings(text: str, findings: Sequence[ManualFinding]) -> None:
        if len(findings) > 1_000:
            raise PipelineInputError("Too many manual findings")
        for finding in findings:
            if finding.entity_type not in ALLOWED_MANUAL_ENTITY_TYPES:
                raise PipelineInputError("Manual finding category is invalid")
            if finding.start < 0 or finding.end <= finding.start or finding.end > len(text):
                raise PipelineInputError("Manual finding offsets are invalid")

    def _detect_all(self, text: str) -> list[Detection]:
        detections: list[Detection] = []
        for detector in self.detectors:
            if detector.status.ready:
                detections.extend(detector.detect(text))
        return [
            detection
            for detection in detections
            if 0 <= detection.start < detection.end <= len(text)
        ]

    @staticmethod
    def _merge_overlaps(detections: Iterable[Detection]) -> list[Detection]:
        ordered = sorted(
            detections,
            key=lambda item: (item.start, item.end, -item.score, item.entity_type),
        )
        if not ordered:
            return []

        merged: list[Detection] = []
        current = ordered[0]
        for candidate in ordered[1:]:
            if candidate.start >= current.end:
                merged.append(current)
                current = candidate
                continue

            current_rank = _detection_rank(current)
            candidate_rank = _detection_rank(candidate)
            winning_type = (
                candidate.entity_type if candidate_rank > current_rank else current.entity_type
            )
            current = Detection(
                start=min(current.start, candidate.start),
                end=max(current.end, candidate.end),
                entity_type=winning_type,
                score=max(current.score, candidate.score),
                recognizers=tuple(sorted(set(current.recognizers + candidate.recognizers))),
                explanation="Combined overlapping local detections",
            )
        merged.append(current)
        return merged

    @staticmethod
    def _replace(text: str, detections: Sequence[Detection]) -> tuple[str, tuple[Replacement, ...]]:
        counters: dict[str, int] = defaultdict(int)
        value_tokens: dict[tuple[str, str], str] = {}
        replacements: list[Replacement] = []

        for detection in detections:
            label = PLACEHOLDER_LABELS[normalize_entity_type(detection.entity_type)]
            transient_value_key = (label, text[detection.start : detection.end].casefold())
            replacement_value = value_tokens.get(transient_value_key)
            if replacement_value is None:
                counters[label] += 1
                replacement_value = f"[{label}_{counters[label]}]"
                value_tokens[transient_value_key] = replacement_value
            replacements.append(
                Replacement(
                    start=detection.start,
                    end=detection.end,
                    value=replacement_value,
                )
            )

        cleaned = text
        for replacement in reversed(replacements):
            cleaned = cleaned[: replacement.start] + replacement.value + cleaned[replacement.end :]
        return cleaned, tuple(replacements)

    @staticmethod
    def _map_original_span(
        start: int,
        end: int,
        replacements: Sequence[Replacement],
    ) -> tuple[int, int] | None:
        delta = 0
        for replacement in replacements:
            if replacement.end <= start:
                delta += len(replacement.value) - (replacement.end - replacement.start)
                continue
            if replacement.start >= end:
                break
            return None
        return start + delta, end + delta

    @classmethod
    def _map_excluded_spans(
        cls,
        *,
        automatic_findings: Sequence[Finding],
        replacements: Sequence[Replacement],
    ) -> tuple[tuple[int, int, str], ...]:
        exemptions: list[tuple[int, int, str]] = []
        for finding in automatic_findings:
            if finding.selected:
                continue
            mapped = cls._map_original_span(finding.start, finding.end, replacements)
            if mapped is not None:
                exemptions.append((mapped[0], mapped[1], finding.entity_type))
        return tuple(exemptions)

    @staticmethod
    def _is_reviewed_exemption(
        detection: Detection,
        exemptions: Sequence[tuple[int, int, str]],
    ) -> bool:
        return any(
            detection.start >= start
            and detection.end <= end
            and detection.entity_type == entity_type
            for start, end, entity_type in exemptions
        )

    @staticmethod
    def _inside_placeholder(text: str, detection: Detection) -> bool:
        return any(
            detection.start >= match.start() and detection.end <= match.end()
            for match in PLACEHOLDER_PATTERN.finditer(text)
        )
